File: app/src/main/java/com/amyllykoski/asyncseq/impl/MyServiceImpl.java
package com.amyllykoski.asyncseq.impl;

import android.os.HandlerThread;
import android.os.MessageQueue;

import com.amyllykoski.asyncseq.api.MyService;
import com.amyllykoski.asyncseq.api.RestCallback;
import com.amyllykoski.asyncseq.model.Item;
import com.amyllykoski.asyncseq.util.L;

import java.util.Date;
import java.util.List;

public class MyServiceImpl implements MyService {
  private static final String TAG = MyServiceImpl.class.getSimpleName();
  private HandlerThread handlerThread;
  private String baseUrl;

  public MyServiceImpl(final String baseUrl) {
    this.baseUrl = baseUrl;
    handlerThread = new HandlerThread(TAG + "-" + new Date().getTime(), HandlerThread.MIN_PRIORITY);
    handlerThread.start();
    setIdleHandler();
  }

  @Override
  public void getItems(final RestCallback<List<Item>> items) {
    ServiceHandler<List<Item>> handler = new ServiceHandler<>(handlerThread.getLooper(), items);
    handler.post(new GetItems(baseUrl, handler));
  }

  @Override
  public void doLoop(long delayMillis, final String tag, final RestCallback<String> response) {
    ServiceHandler<String> handler = new ServiceHandler<>(handlerThread.getLooper(), response);
    handler.post(new DoLoop(delayMillis, tag, handler));
  }

  @Override
  public void close() {
    if (handlerThread != null) {
      handlerThread.quitSafely();
    }
  }

  private void setIdleHandler() {
    MessageQueue.IdleHandler idleHandler = new MessageQueue.IdleHandler() {
      @Override
      public boolean queueIdle() {
        L.d(TAG, "MessageQueue is idle.");
        return true;
      }
    };
    handlerThread.getLooper().getQueue().addIdleHandler(idleHandler);
  }
}

File: app/src/main/java/com/amyllykoski/asyncseq/impl/MyServiceImpl.kt
package com.amyllykoski.asyncseq.impl

import android.os.HandlerThread
import arrow.core.Either
import com.amyllykoski.asyncseq.api.MyError
import com.amyllykoski.asyncseq.api.MyService
import com.amyllykoski.asyncseq.model.Item
import com.amyllykoski.asyncseq.util.MyCallback

class MyServiceImpl(private val baseUrl: String) : MyService {
  private val handlerThread: HandlerThread

  init {
    handlerThread = HandlerThread(TAG, HandlerThread.MIN_PRIORITY)
    handlerThread.start()
  }

  override fun getItems(): Either<MyError, List<Item>> {
    val items = MyCallback<List<Item>>(Companion.TIMEOUT)
    val handler = ServiceHandler(handlerThread.looper, items)
    handler.post(GetItems(baseUrl, handler))
    return items.data
  }

  override fun close() {
    handlerThread.quitSafely()
  }

  companion object {
    private val TAG = MyServiceImpl::class.java.simpleName
    private const val TIMEOUT = 10L
  }
}
